Prose/ProseLib/Sources/App/AppScene.swift

```diff
@@ -0,0 +1,90 @@
+//
+//  AppScene.swift
+//  Prose
+//
+//  Created by Valerian Saliou on 9/14/21.
+//
+
+import AuthenticationFeature
+import ComposableArchitecture
+import MainWindowFeature
+import SettingsFeature
+import SwiftUI
+
+public struct AppScene: Scene {
+    @NSApplicationDelegateAdaptor(AppDelegate.self) var appDelegate
+
+    private let store: Store<AppState, AppAction>
+    private var actions: ViewStore<Void, AppAction> { ViewStore(self.store.stateless) }
+
+    @SwiftUI.State var authWindow: NSWindow?
+    @SwiftUI.State var mainWindow: NSWindow?
+
+    public init(store: Store<AppState, AppAction>) {
+        self.store = store
+    }
+
+    public init() {
+        self.init(store: Store(
+            initialState: AppState(
+                route: .auth(.init())
+            ),
+            reducer: appReducer,
+            environment: AppEnvironment.live
+        ))
+    }
+
+    public var body: some Scene {
+        WithViewStore(self.store.scope(state: \.route.tag)) { viewStore in
+            login()
+                .handlesExternalEvents(matching: Set(arrayLiteral: "login"))
+            main()
+                .handlesExternalEvents(matching: Set(arrayLiteral: "main"))
+                // NOTE: [Rémi Bardon] Make sure `.onChange` is called only once.
```

Comment:
Isn't that satisfied now?

Reply:
I wasn't clear

I meant "I put it on `main()`, even though it relates to `auth()` too, but if you put it on `Group { auth(); main() } then it'll be run twice`"

Reply:
I see. And is that bad? That should be a no-op?

Which of course begs the question why even scope then, etc. Yet here we are. But calling it twice shouldn't be a bad thing and we can remove the comment and leave this thing as is. Or do you have a different view?

Reply:
I can remove the comment, it just felt a bit weird putting the modifier on only one view. Sometimes SwiftUI doesn't run `.onChange`s for views not on screen, but hopefully an empty `WindowGroup` doesn't behave like `EmptyView`.

> Which of course begs the question why even scope then, etc. Yet here we are.

What do you mean?

Reply:
And if you put it on a `Group` as you've suggested? It's run twice then, but that seems ok to me.

Reply:
I don't really like it being run twice either… I'd like someting like `.drawingGroup`, which really groups things together instead of keeping them individual views, but I don't think it's possible for `Scene`s.

Reply:
Btw, it was a `NOTE`, and not a `TODO`

Reply:
A `NOTE` saying "Make sure…" sounds like `TODO` to me 😬.

So why not run it twice making sure that …

> Sometimes SwiftUI doesn't run .onChanges for views not on screen

doesn't happen? What are you trying to prevent?

Reply:
I did

```swift
Group {
    login()
        .handlesExternalEvents(matching: Set(arrayLiteral: "login"))
    main()
        .handlesExternalEvents(matching: Set(arrayLiteral: "main"))
}
// NOTE: [Rémi Bardon] This `.onChange` is called multiple times, but at least we're sure it runs.
//       See <https://github.com/prose-im/prose-app-macos/pull/22> for discussion about it.
.onChange(of: viewStore.state) { newRoute in
    switch newRoute {
    case .auth:
        mainWindow?.close()
    case .main:
        authWindow?.close()
    }
}
```